src/tFPDF/ProtectedPDF.php

```diff
@@ -0,0 +1,271 @@
+    const PERMISSION_PRINT = 4;
+    const PERMISSION_MODIFY = 8;
+    const PERMISSION_COPY = 16;
+    const PERMISSION_ANNOTATE_FORMS = 32;
+    const PROTECTION_BASE = 192;
+    const PASSWORD_LENGTH = 32;
+
+    /**
+     * Whether encryption is enabled
+     *
+     * @var bool
+     */
+    private $bol_encrypted = false;
+
+    /**
+     * @var int
+     */
+    private $int_encryption_object_id;
+
+    /**
+     * The generated encryption key for the document
+     *
+     * @var string
+     */
+    private $str_encryption_key;
+
+    /**
+     * @var string
+     */
+    private $str_u_value;
+
+    /**
+     * @var string
+     */
+    private $str_o_value;
+
+    /**
+     * @var string
+     */
+    private $int_p_value;
+
+    /** {@inheritdoc} */
+    protected function TextString($s)
+    {
+        if ($this->bol_encrypted) {
+            // Encrypt text
+            $s = self::RC4($this->ObjectKey($this->int_current_object), $s);
+        }
+
+        return parent::TextString($s);
+    }
+
+    /** {@inheritdoc} */
+    protected function PutStream($str_data)
+    {
+        if ($this->bol_encrypted) {
+            // Encrypt content
+            $str_data = self::RC4($this->ObjectKey($this->int_current_object), $str_data);
+        }
+
+        parent::PutStream($str_data);
+    }
+
+    /** {@inheritdoc} */
+    public function PutResources()
+    {
+        parent::PutResources();
+
+        if ($this->bol_encrypted) {
+            $this->NewObject();
+            $this->int_encryption_object_id = $this->int_current_object;
+            $this->Out('<<');
+            $this->PutEncryption();
+            $this->Out('>>');
+            $this->Out('endobj');
+        }
+    }
+
+    /** {@inheritdoc} */
+    public function PutTrailer()
+    {
+        parent::PutTrailer();
+
+        if ($this->bol_encrypted) {
+            $this->Out('/Encrypt ' . $this->int_encryption_object_id . ' 0 R');
+            $this->Out('/ID [()()]');
+        }
+    }
+
+    /**
+     * Write encryption metadata to the document
+     */
+    public function PutEncryption()
+    {
+        $this->Out('/Filter /Standard');
+        $this->Out('/V 1');
+        $this->Out('/R 2');
+
+        $this->Out('/O ('.$this->EscapeString($this->str_o_value).')');
+        $this->Out('/U ('.$this->EscapeString($this->str_u_value).')');
+        $this->Out('/P '.$this->int_p_value);
+    }
+
+    /**
+     * @param array|null $permissions An array of ::PERMISSION_*
+     * @param string $userPassword The password for restricted access to the document (obeying $permissions)
+     * @param string|null $ownerPassword The password for unrestricted access to the document (ignoring $permissions)
+     */
+    public function SetProtection(array $permissions = null, $userPassword = '', $ownerPassword = null)
+    {
+        $protection = self::PROTECTION_BASE;
+        foreach((array)$permissions as $permission){
+            $protection += $permission;
+        }
```

Comment:
I think it would be better to do a binary check before adding each permission value. Just to prevent issues where the same flag is enabled twice, e.g.
```php
$pdf_instance->setProtection([$pdf_instance::PERMISSION_PRINT, $pdf_instance::PERMISSION_PRINT]);
```
Will actually add print twice, resulting in PERMISSION_MODIFY being enabled instead (but not print!).

I know this sounds like the developers is asking for trouble himself when passing duplicate values, but it's probably better to be safe than sorry, especially when dealing with such delicate things as permissions.

What I would suggest is:
```php
$protection = self::PROTECTION_BASE;
foreach ((array)$permissions as $permission){
    if (($protection & $permission) == 0) {
        $protection += $permission;
    }
}
```